librubyfmt/src/format.rs

```diff
@@ -2807,287 +2744,155 @@ fn can_elide_parens_for_reserved_names(cc: &[CallChainElement]) -> bool {
 fn format_call_chain(
     ps: &mut dyn ConcreteParserState,
     cc: Vec<CallChainElement>,
-    should_multiline_call_chain: bool,
-) -> bool {
+    last_call_use_parens: Option<bool>,
+) {
     if cc.is_empty() {
-        return false;
+        return;
     }
 
-    format_call_chain_elements(ps, cc, should_multiline_call_chain);
+    let first_elem_line = cc.first().unwrap().start_line();
+    if let Some(first_elem_line) = first_elem_line {
+        ps.on_line(first_elem_line);
+    }
+
+    ps.breakable_call_chain_of(
+        cc.clone(),
+        Box::new(|ps| format_call_chain_elements(ps, cc, last_call_use_parens)),
+    );
 
     ps.emit_after_call_chain();
-    should_multiline_call_chain
 }
 
 fn format_call_chain_elements(
     ps: &mut dyn ConcreteParserState,
     cc: Vec<CallChainElement>,
-    render_multiline_chain: bool,
+    // Whether or not to force the last call to use parens. By default, falls back to normal call chain rules.
+    // This is necessary for supporting things like parenthesized methods in `self.method` method chains where
+    // the parens are sometimes semantically meaningful. However, we leave this as optional since not all callers
+    // require this (e.g. `MethodAddArg` doesn't enforce invariants like those).
+    last_call_use_parens: Option<bool>,
 ) {
     let elide_parens = can_elide_parens_for_reserved_names(&cc);
-    let mut has_indented = false;
     // When set, force all `CallChainElement::ArgsAddStarOrExpressionListOrArgsForward`
     // to use parens, even when empty. This handles cases like `super()` where parens matter
     let mut next_args_list_must_use_parens = false;
-    for cc_elem in cc {
-        let mut element_is_super_keyword = false;
+    let last_call_index = cc
+        .iter()
+        .rposition(|cce| matches!(cce, CallChainElement::IdentOrOpOrKeywordOrConst(..)));
+    let mut has_indented = false;
+
+    for (index, cc_elem) in cc.into_iter().enumerate() {
+        let mut element_must_use_parens = false;
+        let is_last_call_args = if let Some(last_call_index) = last_call_index {
+            index == (last_call_index + 1)
+        } else {
+            false
+        };
 
         match cc_elem {
             CallChainElement::Paren(p) => format_paren(ps, p),
             CallChainElement::IdentOrOpOrKeywordOrConst(i) => {
                 let ident = i.into_ident();
-                element_is_super_keyword = ident.1 == "super";
+                element_must_use_parens = ident.1 == "super";
 
-                format_ident(ps, ident)
+                if ident.1 == ".()" {
+                    ps.emit_ident(".".to_string());
+                    element_must_use_parens = true;
+                } else {
+                    format_ident(ps, ident);
+                }
+                ps.shift_comments();
             }
             CallChainElement::Block(b) => {
                 ps.emit_space();
                 format_block(ps, b)
```

Comment:
are we missing a shift_comments call here?

Reply:
I don't think we need to because `format_{do,brace}_block` should handle comment shifting internally (since they'll want to render the comments inside the block delimiters), but I'll dd a comment here to make that explicit